File: macos/Sources/ElevateCore/Providers/PolicyRules.swift
import Foundation

/// The end-user rules of a PIM role management policy. Graph (Entra, groups) and ARM (Azure resources)
/// wrap them in different envelopes but the rules themselves — and how they map to a `RolePolicy` — are identical.
public enum PolicyRules {
    public struct Rule: Decodable {
        let id: String
        let isExpirationRequired: Bool?
        let maximumDuration: String?
        let enabledRules: [String]?
        let setting: ApprovalSetting?
        let isEnabled: Bool?
        let claimValue: String?
        public struct ApprovalSetting: Decodable { let isApprovalRequired: Bool? }
    }

    /// Folds the end-user rules onto `.manualDefault`; rules we do not understand are ignored.
    public static func apply(_ rules: [Rule]) -> RolePolicy {
        var policy = RolePolicy.manualDefault
        for rule in rules {
            switch rule.id {
            case "Expiration_EndUser_Assignment":
                if let d = rule.maximumDuration.flatMap(ISO8601Duration.parse) {
                    policy.maximumDuration = d
                    policy.defaultDuration = d
                }
            case "Enablement_EndUser_Assignment":
                let enabled = Set(rule.enabledRules ?? [])
                policy.requiresJustification = enabled.contains("Justification")
                policy.requiresTicket = enabled.contains("Ticketing")
                policy.requiresMFA = enabled.contains("MultiFactorAuthentication")
            case "Approval_EndUser_Assignment":
                policy.requiresApproval = rule.setting?.isApprovalRequired ?? false
            case "AuthenticationContext_EndUser_Assignment":
                if rule.isEnabled == true, let claim = rule.claimValue, !claim.isEmpty { policy.authenticationContext = claim }
            default:
                break
            }
        }
        return policy
    }
}

File: windows/src/Elevate.Core/Providers/PolicyRules.cs
using Elevate.Core.Models;
using Elevate.Core.Support;

namespace Elevate.Core.Providers;

/// <summary>
/// The end-user rules of a PIM role management policy. Graph (Entra, groups) and ARM (Azure resources)
/// wrap them in different envelopes but the rules themselves — and how they map to a
/// <see cref="RolePolicy"/> — are identical.
/// </summary>
public static class PolicyRules
{
    public sealed record Rule(
        string Id,
        bool? IsExpirationRequired,
        string? MaximumDuration,
        IReadOnlyList<string>? EnabledRules,
        Rule.ApprovalSetting? Setting,
        bool? IsEnabled,
        string? ClaimValue)
    {
        public sealed record ApprovalSetting(bool? IsApprovalRequired);
    }

    /// <summary>Folds the end-user rules onto <see cref="RolePolicy.ManualDefault"/>; unknown rules are ignored.</summary>
    public static RolePolicy Apply(IEnumerable<Rule> rules)
    {
        ArgumentNullException.ThrowIfNull(rules);

        var policy = RolePolicy.ManualDefault;
        foreach (var rule in rules)
        {
            switch (rule.Id)
            {
                case "Expiration_EndUser_Assignment":
                    if (rule.MaximumDuration is { } text && Iso8601Duration.Parse(text) is { } duration)
                    {
                        policy = policy with { MaximumDuration = duration, DefaultDuration = duration };
                    }

                    break;
                case "Enablement_EndUser_Assignment":
                {
                    var enabled = new HashSet<string>(rule.EnabledRules ?? [], StringComparer.Ordinal);
                    policy = policy with
                    {
                        RequiresJustification = enabled.Contains("Justification"),
                        RequiresTicket = enabled.Contains("Ticketing"),
                        RequiresMfa = enabled.Contains("MultiFactorAuthentication"),
                    };
                    break;
                }

                case "Approval_EndUser_Assignment":
                    policy = policy with { RequiresApproval = rule.Setting?.IsApprovalRequired ?? false };
                    break;
                case "AuthenticationContext_EndUser_Assignment":
                    if (rule.IsEnabled == true && rule.ClaimValue is { Length: > 0 } claim)
                    {
                        policy = policy with { AuthenticationContext = claim };
                    }

                    break;
                default:
                    break;
            }
        }

        return policy;
    }
}
